Restructure SCF

ndaelman-hu opened the pull request:
Specialize convergence reporting to dedicated properties (`ConvergenceMetrics`), instead of the general `PhysicalProperty`.
Define convergence plots based on `ConvergenceMetrics`. Fall back to inheritance for unit specification.

ndaelman-hu committed:
Demote `value` to status of suggested name, not default

ndaelman-hu committed:
Remove mention to `variables`

ndaelman-hu committed:
- Reshape approach to SCF:
-- Remove `is_scf_converged` and the corresponding references from generic `PhysicalProperty`
-- Move SCF output to specialized base section `ConvergenceMetric`
-- Specify plotting functionaltities in `ConvergenceMetric`
-- Declare SCF nature in methods

ndaelman-hu committed:
- Reshape approach to SCF:
-- Declare SCF nature in methods

ndaelman-hu committed:
- Add `is_scf_converged` to `ConvergenceMetric`
- Specialize `SelfConsistency` and `ConvergenceMetrics`, e.g. energies, forces
- TODO: connect `SelfConsistency` with `ConvergenceMetrics`

ndaelman-hu commented:
@ladinesa Do you think that connecting `SelfConsistency` to `ConvergenceMetrics` is appropriate via `after_normalize` (i.e. the parser), or is `normalize` better?

@JFRudzinski I know that we have this division between inputs (system and method) and outputs, but in the case of convergences, keeping them separate just needlessly requires logic to enforce connection. Moreover, there are many kinds of energy differences. 
@ladinesa How feasible would it be all quantities in the same section (both input and output running over the scf cycles) via the mapping annotation parser?

ndaelman-hu committed:
- Add reminder for `name` as semantic distinghuisher
- Apply ruff

ndaelman-hu committed:
Correct missing attribute

ladinesa commented:
> @ladinesa Do you think that connecting `SelfConsistency` to `ConvergenceMetrics` is appropriate via `after_normalize` (i.e. the parser), or is `normalize` better?
> 

I would do it at the root section normalize method.


> @JFRudzinski I know that we have this division between inputs (system and method) and outputs, but in the case of convergences, keeping them separate just needlessly requires logic to enforce connection. Moreover, there are many kinds of energy differences. @ladinesa How feasible would it be all quantities in the same section (both input and output running over the scf cycles) via the mapping annotation parser?

You can simply return a list of the scf cycles and mapping parser will run through the list of all quantities in the section with mapping annotation. what you cannot map are the references.

EBB2675 commented:
I don't think I could follow it. 

Just importing `EnergySelfConsistency` already crashes, proving the new class still relies on a `differential_threshold` quantity that isn’t defined.

Do I understand correctly that when 

`differential_threshold = SelfConsistency.m_def.differential_threshold.m_copy()`

runs, Python is still building `SelfConsistency`; its metainfo (`SelfConsistency.m_def`) isn’t populated yet, hence `AttributeError`?

EBB2675 commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -861,23 +855,22 @@ class SelfConsistency(NumericalSettings):
     )
 
-    threshold_change = Quantity(
+    differential_threshold = Quantity(
```
`differential_threshold` feels off because **“differential” ≈ derivative** in math, while here you just mean change between steps. Better options:

- `convergence_tolerance` :  common SCF term 

- `delta_threshold` or `change_threshold` : explicit that it’s the allowed Δ between iterations.

- `step_tolerance` : highlights it’s measured per step.

 If you need the unit-specific subclasses:

 `EnergySelfConsistency.energy_tolerance`, `ForceSelfConsistency.force_tolerance`.

We have a lot more of these thresholds & tolerances, it might be better if we agree on a common ground ( i do not have any preferences )

ndaelman-hu replied:
I'll retain the original `change_threshold` then, though `tolerance` has my personal preference.

ndaelman-hu replied:
> We have a lot more of these thresholds & tolerances, it might be better if we agree on a common ground ( i do not have any preferences )

What do you mean here? _Different thresholds_, as in _different properties_ being tracked, or _different ways of tracking_?

EBB2675 replied:
both, hence I didn't distinguish

ndaelman-hu replied:
Okay. The former (_different properties_) is already somewhat captured by the sections' polymorphism, e.g. `EnergySelfConsistency` and `ForceSelfConsistency`. It isn't meant to be an exhaustive list, though strain could be a standard addition.

This approach does not distinguish certain classes of energy measurement, e.g. total energy change, [residuum change](https://www.vasp.at/wiki/index.php/OSZICAR).
Those can either be captured with more inheritance, or a quantity like `label`. I prefer `label`, as I want to avoid extensions to the schema as much as possible.

I couldn't immediately come up with examples of the latter (_different ways of tracking_). @EBB2675 could you list some alternatives? Feel free to also weigh in on how to integrate them into the schema.

EBB2675 replied:
Sure—by “different ways of tracking” I just meant the various formulas codes use for the per-step change:

- **Absolute change** ‖Eₙ − Eₙ₋₁‖, ‖Fₙ − Fₙ₋₁‖

- **Relative change** ‖ΔE / Eₙ‖, ‖Δρ / ρₙ‖

- **Max component** max|Δρ(r)|, max|ΔF_i|

- **RMS / L² norm** √(Σ Δx² / N), DIIS error vector norm

 - **Residuum / gradient** ‖[F,P]‖, ‖∇E‖

 - **Compound criteria** e.g. VASP converges when both ΔE and Δρ are below their thresholds.

ndaelman-hu replied:
Thx! Except for _compound criteria_, I'll capture this with an `Enum` quantity.

For compounds, if we assume that convergence may generally be written as `(requirement_1 && requirement_2 && ...) || (... && ...) || ...`, then we need tags for `&&` groups. This is very complex, as the mapping annotations work asynchronously from our POV.
Limiting the convergence requirements to a single `&&` would be much simpler, as a single quantity suffices.

ndaelman-hu commented:
> I don't think I could follow it.
> 
> Just importing `EnergySelfConsistency` already crashes, proving the new class still relies on a `differential_threshold` quantity that isn’t defined.
> 
> Do I understand correctly that when
> 
> `differential_threshold = SelfConsistency.m_def.differential_threshold.m_copy()`
> 
> runs, Python is still building `SelfConsistency`; its metainfo (`SelfConsistency.m_def`) isn’t populated yet, hence `AttributeError`?

Okay, will take a look at the breaking attribute. Thx.

Overall, I need my question regarding design resolved first, before I fix testing and the like.
This was just an early stage review request. If you want, you can skip it, and I'll let you know once the tests pass.

EBB2675 reviewed (commented):
i will get back to this review once it is finished

ndaelman-hu commented:
> > @ladinesa Do you think that connecting `SelfConsistency` to `ConvergenceMetrics` is appropriate via `after_normalize` (i.e. the parser), or is `normalize` better?
> 
> I would do it at the root section normalize method.
> 
> > @JFRudzinski I know that we have this division between inputs (system and method) and outputs, but in the case of convergences, keeping them separate just needlessly requires logic to enforce connection. Moreover, there are many kinds of energy differences. @ladinesa How feasible would it be all quantities in the same section (both input and output running over the scf cycles) via the mapping annotation parser?
> 
> You can simply return a list of the scf cycles and mapping parser will run through the list of all quantities in the section with mapping annotation. what you cannot map are the references.

Indeed, mapping annotation cannot handle the references. I can make a workaround, that registers every setting for quick look-up.
Or, we avoid references by having inputs and outputs together, in a single section. The output convergence is a single, array-like quantity. These are typically dispersed over the output file. I was curious whether this could pose any trouble for mapping annotation.

ndaelman-hu committed:
- Change `differential_threshold` to `convergence_tolerance`
- Correct schema copying for `EnergySelfConsistency` and `ForceSelfConsistency`

ndaelman-hu committed:
Add `comparison_mode`

ndaelman-hu committed:
Add `is_required`

JFRudzinski commented:
Closing in favor of #260 

@ndaelman-hu
